lib/core/src/Cardano/Wallet/Primitive/Slotting.hs

```diff
@@ -201,124 +202,140 @@ ongoingSlotAt x = do
+    Range (Just Nothing) _ -> Nothing
+    Range _ (Just Nothing) -> Nothing
+    Range a b -> Just (Range (join a) (join b))
+
+-- | Transforms the given inclusive time range into an inclusive slot range.
+slotRangeFromRelativeTimeRange :: Range RelativeTime -> Qry (Range SlotNo)
+slotRangeFromRelativeTimeRange (Range a b) =
+    Range <$> traverse ceilingSlotAt a <*> traverse ongoingSlotAt b
+
+slotRangeFromTimeRange :: Range UTCTime -> Qry (Maybe (Range SlotNo))
+slotRangeFromTimeRange range = mapM slotRangeFromRelativeTimeRange
+    =<< (toRelativeTimeRange range <$> getStartTime)
+
+{-------------------------------------------------------------------------------
+                            Blockchain-relative time
+-------------------------------------------------------------------------------}
+
+-- | Same as 'Cardano.toRelativeTime', but has error handling for times before
+-- the system start. No other functions in this module will accept UTC times.
+toRelativeTime :: StartTime -> UTCTime -> Maybe RelativeTime
+toRelativeTime (StartTime start) utc
+    | utc < start = Nothing
+    | otherwise = Just $ Cardano.toRelativeTime (SystemStart start) utc
+
+-- | Convert an absolute time to a relative time. If the absolute time is before
+-- the system start, consider the relative time to be the system start
+-- time. This function can never fail.
+toRelativeTimeOrZero :: StartTime -> UTCTime -> RelativeTime
+toRelativeTimeOrZero start = fromMaybe (RelativeTime 0) . toRelativeTime start
+
+-- | Query the absolute time corresponding to a blockchain-relative time.
+fromRelativeTime :: RelativeTime -> Qry UTCTime
+fromRelativeTime t = do
+    start <- getStartTime
+    pure (Cardano.fromRelativeTime (coerce start) t)
+
+{-------------------------------------------------------------------------------
+                                What's the time?
+-------------------------------------------------------------------------------}
+
+-- | The current system time, compared to the given blockchain start time.
+--
+-- If the current time is before the system start (this would only happen when
+-- launching testnets), let's just say we're in epoch 0.
 --
--- NOTE:
--- A @TimeInterpreter@ could in theory decide to update the era summary from the
--- node when running a query.
+-- TODO: Use io-sim-classes for easier testing.
+getCurrentTimeRelativeFromStart :: MonadIO m => StartTime -> m RelativeTime
+getCurrentTimeRelativeFromStart start =
+    toRelativeTimeOrZero start <$> liftIO getCurrentTime
+
+-- | The current system time, compared to the blockchain start time of the given
+-- 'TimeInterpreter'.
 --
--- We cannot manually specify when the fetching happens.
+-- If the current time is before the system start (this would only happen when
+-- launching testnets), the relative time is reported as 0.
+currentRelativeTime :: MonadIO m => TimeInterpreter m -> m RelativeTime
+currentRelativeTime ti = ti getStartTime >>= getCurrentTimeRelativeFromStart
```

Comment:
Rebasing #2245 I found that if `m ~ ExceptT PastHorizonException IO` then suddenly the `currentRelativeTime` can fail… with `PastHorizonException`, which doesn't make sense.

I changed it to
```haskell
currentRelativeTime :: MonadIO m => TimeInterpreter n -> m RelativeTime
```
but we can deal with that in #2245.

Reply:
OK thanks.
Maybe we should change `MonadIO` to just `IO`, or `MonadTime` (from io-sim-classes package).